File: libs/power-log-parser/src/lib/parsers/winner-parser.ts
import { GameTag, PlayState } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventProvider } from '../game-event';
import { Node, PlayerEntity, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class WinnerParser implements ActionParser {
	readonly ParserName = 'WinnerParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === TagChange &&
			((node.Object as TagChange).Name === (GameTag.PLAYSTATE as number) ||
				(this.ParserState.IsBattlegrounds() &&
					(node.Object as TagChange).Name ===
						(GameTag.TAG_PLAYER_CONCEDED_OR_DISCONNECTED as number)))
		);
	}

	AppliesOnCloseNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const tagChange = node.Object as TagChange;
		if (
			tagChange.Name === (GameTag.PLAYSTATE as number) &&
			tagChange.Value === (PlayState.WON as number)
		) {
			const winner = this.ParserState.GetEntity(tagChange.Entity) as PlayerEntity | undefined;
			if (winner == null) {
				return null;
			}

			return [
				GameEventProvider.Create(
					tagChange.TimeStamp,
					'WINNER',
					() => ({
						Type: 'WINNER',
						Value: {
							Winner: winner,
							LocalPlayer: this.StateFacade.LocalPlayer,
							OpponentPlayer: this.StateFacade.OpponentPlayer,
						},
					}),
					true,
					node,
				),
			];
		} else if (
			tagChange.Name === (GameTag.PLAYSTATE as number) &&
			tagChange.Value === (PlayState.TIED as number)
		) {
			return [
				GameEventProvider.Create(
					tagChange.TimeStamp,
					'TIE',
					() => ({
						Type: 'TIE',
					}),
					true,
					node,
				),
			];
		} else if (tagChange.Name === (GameTag.TAG_PLAYER_CONCEDED_OR_DISCONNECTED as number)) {
			const isPlayer = tagChange.Entity === this.StateFacade.LocalPlayer?.Id;
			if (!isPlayer) {
				return null;
			}
			return [
				GameEventProvider.Create(
					tagChange.TimeStamp,
					'WINNER',
					() => ({
						Type: 'WINNER',
						Value: {
							Winner: this.StateFacade.OpponentPlayer,
							LocalPlayer: this.StateFacade.LocalPlayer,
							OpponentPlayer: this.StateFacade.OpponentPlayer,
						},
					}),
					true,
					node,
				),
			];
		}
		return null;
	}

	CreateGameEventProviderFromClose(_node: Node): GameEventProvider[] | null {
		return null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/WinnerParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;

namespace HearthstoneReplays.Events.Parsers
{
    public class WinnerParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public WinnerParser(ParserState ParserState, StateFacade helper)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = helper;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(TagChange)
                && ((node.Object as TagChange).Name == (int)GameTag.PLAYSTATE 
                    || (ParserState.IsBattlegrounds() && (node.Object as TagChange).Name == (int)GameTag.TAG_PLAYER_CONCEDED_OR_DISCONNECTED));
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return false;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var tagChange = node.Object as TagChange;
            if (tagChange.Name == (int)GameTag.PLAYSTATE && tagChange.Value == (int)PlayState.WON)
            {
                var winner = (PlayerEntity)ParserState.GetEntity(tagChange.Entity);
                if (winner == null)
                {
                    return null;
                }

                //Logger.Log("Creating event provider for WinnerParser", node.CreationLogLine);
                return new List<GameEventProvider> { GameEventProvider.Create(
                       tagChange.TimeStamp,
                       "WINNER",
                       () => {
                            //Logger.Log("Providing game event for WinnerParser", node.CreationLogLine);
                            return new GameEvent
                            {
                                Type = "WINNER",
                                Value = new
                                {
                                    Winner = winner,
                                    LocalPlayer = StateFacade.LocalPlayer,
                                    OpponentPlayer = StateFacade.OpponentPlayer,
                                }
                            };
                       },
                       true,
                       node) };
            }
            else if (tagChange.Name == (int)GameTag.PLAYSTATE && tagChange.Value == (int)PlayState.TIED)
            {
                return new List<GameEventProvider> { GameEventProvider.Create(
                       tagChange.TimeStamp,
                        "TIE",
                       () => new GameEvent
                       {
                           Type = "TIE"
                       },
                       true,
                       node) };
            }
            else if (tagChange.Name == (int)GameTag.TAG_PLAYER_CONCEDED_OR_DISCONNECTED)
            {
                var isPlayer = tagChange.Entity == StateFacade.LocalPlayer.Id;
                if (!isPlayer)
                {
                    return null;
                }
                return new List<GameEventProvider> { GameEventProvider.Create(
                       tagChange.TimeStamp,
                       "WINNER",
                       () => {
                            //Logger.Log("Providing game event for WinnerParser", node.CreationLogLine);
                            return new GameEvent
                            {
                                Type = "WINNER",
                                Value = new
                                {
                                    Winner = StateFacade.OpponentPlayer,
                                    LocalPlayer = StateFacade.LocalPlayer,
                                    OpponentPlayer = StateFacade.OpponentPlayer,
                                }
                            };
                       },
                       true,
                       node) };

            }
            return null;
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            return null;
        }
    }
}
